### PR TITLE
[Draft] Angular Apps support

### PR DESCRIPTION
**Pull request by nvladimirovi**
**Angular Isolation**
This PR gives the user the ability to create UI Plugins with Angular and Clarity isolation.

**Window Object Isolation**
Using webpack DefinePlugin the user will be able to isolate the window object as well.

**CSS Isolation**
Using our PrecalculateRem postcss plugin the user will be able to isolate the styles of the plugin in combination with Shadow DOM.

**Breaking Change**
After this change plugin developers will have to disable ivy using tsconfig.json or use angular.json 
Example:
// tsconfig.json
...
.
"angularCompilerOptions": {
        "enableIvy": false,
 },

// angular.json
"builder": "@vcd/plugin-builders:plugin-builder",
          "options": {
           "modulePath": "src/main/your.plugin.module.ts#YourModule",
           "forceDisableIvy": true
...
.

### CODE REVIEW

File: packages/angular/projects/vcd/plugin-builders/src/lib/base/index.ts [hunk 1]
```diff
@@ -60,13 +69,17 @@ async function commandBuilder(
     const pluginLibsBundles = new Map<string, string>();
 
     // Get the configuration
-    const config = configs.config[0];
+    const config = configs.config[0] || configs.config as webpack.Configuration;
 
     // Make sure we are producing a single bundle
-    delete config.entry.polyfills;
+    delete config.entry['polyfills'];
```
Comment (irahov) [inline packages/angular/projects/vcd/plugin-builders/src/lib/base/index.ts:80]: Here and all other similar changes:

Isn't it possible to preserve the '.' property access?
It is safer than the '[prop_name]' one

File: packages/angular/projects/vcd/plugin-builders/src/lib/base/index.ts [hunk 2]
```diff
@@ -120,7 +134,7 @@ async function commandBuilder(
     }
 
     // Define amd lib
-    config.output.filename = 'bundle.js';
+    config.output.filename = '[name].js';
```
Comment (irahov) [inline packages/angular/projects/vcd/plugin-builders/src/lib/base/index.ts]: The previous code change was exactly aimed at using 'bundle.js'
Why it is needed to revert it back to '[name].js'?

File: packages/angular/projects/vcd/plugin-builders/src/lib/common/interfaces.ts [hunk 3]
```diff
@@ -183,4 +185,53 @@ export interface BasePluginBuilderSchema {
...
+     * Example:
+     * propList: ['font-size']
+     * 
+     * will precalculate the rem only of the font-size css.
+     * 
+     * Defaults to: ["*"]
+     */
+    propList?: string[],
+    /**
+     * Replace given css property which subject to rem precalculations,
```
Comment (irahov) [inline packages/angular/projects/vcd/plugin-builders/src/lib/common/interfaces.ts:240]: Not clear

File: packages/angular/projects/vcd/plugin-builders/src/lib/common/postcss-precalculate-rem.ts [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+            if (opts.replace) {
+                decl.value = value;
+            } else {
+                decl.parent.insertAfter(i, decl.clone({ value: value }));
+            }
+        });
+
+        if (opts.mediaQuery) {
+            css.walkAtRules('media', function (rule) {
+                if (rule.params.indexOf('rem') === -1) return;
```
Comment (irahov) [inline packages/angular/projects/vcd/plugin-builders/src/lib/common/postcss-precalculate-rem.ts]: Better keep the common syntax with '{'  :
if() {
   return ...
}

**Review by irahov — APPROVED** [review]
Master Nick :)
What about testing - how is this project been verified?

File: packages/angular/projects/vcd/plugin-builders/src/lib/base/index.ts (hunk 1 above)
Comment (nvladimirovi) [reply packages/angular/projects/vcd/plugin-builders/src/lib/base/index.ts:80]: The problem is that in webpack `entry?: string | string[] | Entry | EntryFunc;` where
`interface Entry {
        [name: string]: string | string[];
    }`
and therefore dot syntax doesn't work.

File: packages/angular/projects/vcd/plugin-builders/src/lib/common/interfaces.ts (hunk 3 above)
Comment (nvladimirovi) [reply packages/angular/projects/vcd/plugin-builders/src/lib/common/interfaces.ts:240]: New comment

> /**
>      * Replace given css property that is subject to rem precalculations,
>      * if set to true replaces all properties from the `propList` with
>      * their alternative.
>      * 
>      * Example:
>      * .css_class {
>      *      font-size: 1rem;
>      * }
>      * 
>      * if replace is True
>      * .css_class {
>      *      font-size: calc(--css-var-plugin-rem-scaled * 1rem);
>      * }
>      * 
>      * if replace is False
>      * .css_class {
>      *      font-size: 1rem;
>      *      font-size: calc(--css-var-plugin-rem-scaled * 1rem);
>      * }
>      *
>      * Defaults to: true
>      */

File: packages/angular/projects/vcd/plugin-builders/src/lib/common/postcss-precalculate-rem.ts (hunk 4 above)
Comment (nvladimirovi) [reply packages/angular/projects/vcd/plugin-builders/src/lib/common/postcss-precalculate-rem.ts]: fixed in commit -> [lint: fix npm run lint](https://github.com/vmware/vcd-ext-sdk/pull/234/commits/e3b1fd23aaa9519c4cdb9aa88686213f01435ab7)

**Comment by nvladimirovi** [comment]
> Master Nick :)

Master Nick - that's extra honor, am I not still a samurai anymore? 😆 

> What about testing - how is this project been verified?

We test it indirectly with our internal VMware Plugins, but we can think of a better way and place it in this repo.

File: packages/angular/projects/vcd/plugin-builders/src/lib/common/interfaces.ts (hunk 3 above)
Comment (nvladimirovi) [reply packages/angular/projects/vcd/plugin-builders/src/lib/common/interfaces.ts:240]: This discussion will be resolved later.

**Pull request merged**